### PR TITLE
`Programming exercises`: Synchronize repository locks after exercise update

### PR DESCRIPTION
**Pull request by JohannesStoehr**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context

Sometimes instructors have to change certain settings for programming exercises during their conduction. Some of these changes might need locking or unlocking the student repositories. This is now implemented.

### Description

The repositories should be **unlocked** if:
- the due date is postponed and was previously before now
- the offline IDE gets allowed or 
- the exercise gets released by changing the release date

They should be **locked** if:
- the due date is set in the past and was previously not
- if the offline IDE gets forbidden or
- when the exercise gets unreleased by changing the release date

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Programming Exercise

1. Log in to Artemis
2. Participate as a student
3. Do one of the mentioned changes and check if the repository got locked/unlocked correctly (Go to the repository settings in BitBucket as instructor and check the access rights of the student). The changes might not happen instantly since they have to be broadcasted through the InstanceMessegingService.
4. Check the other possible changes

#### Exam Mode Testing

Prerequisites:
- 1 Instructor
- 1 Students
- 1 Exam with a Programming Exercise

1. Log in to Artemis
2. Participate in the exam as a student
3. Check that allowing/forbidding the offline IDE works here as well

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Exam Mode Test
- [x] Test

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|
| ProgrammingExerciseService.java | 82% | 92% |

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java [hunk 1]
```diff
@@ -1073,4 +1073,34 @@ public void deleteTasksWithSolutionEntries(Long exerciseId) {
...
+
+        if (updatedProgrammingExercise.isAllowOfflineIde()) {
+            if (programmingExerciseBeforeUpdate.getDueDate() != null && programmingExerciseBeforeUpdate.getDueDate().isBefore(now)
+                    && (updatedProgrammingExercise.getDueDate() == null || updatedProgrammingExercise.getDueDate().isAfter(now))) {
+                // New due date allows students to continue working on exercise
+                unlockAllRepositories(programmingExerciseBeforeUpdate.getId());
+            }
+            else if ((programmingExerciseBeforeUpdate.getDueDate() == null || programmingExerciseBeforeUpdate.getDueDate().isAfter(now))
+                    && updatedProgrammingExercise.getDueDate() != null && updatedProgrammingExercise.getDueDate().isBefore(now)) {
+                // New due date forbids students to continue working on exercise
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java:1101]: I think lockAllRepositories does not look into the individual due date, but this should get respected in cases like this,

Comment (JohannesStoehr) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java]: Handled now

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java [hunk 2]
```diff
@@ -1073,4 +1073,67 @@ public void deleteTasksWithSolutionEntries(Long exerciseId) {
...
+                // Hide exercise again and lock repos
+                lockAllRepositories(programmingExerciseBeforeUpdate.getId());
+            }
+            return;
+        }
+
+        boolean lockedUnlockedRepos = handleRepoAccessRightChangesDueDates(programmingExerciseBeforeUpdate, updatedProgrammingExercise);
+        if (lockedUnlockedRepos) {
+            return;
+        }
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java:1101]:
```suggestion
        if (handleRepoAccessRightChangesDueDates(programmingExerciseBeforeUpdate, updatedProgrammingExercise) {
            return;
        }
```

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java [hunk 3]
```diff
@@ -1073,4 +1073,67 @@ public void deleteTasksWithSolutionEntries(Long exerciseId) {
...
+    }
+
+    private boolean handleRepoAccessRightChangesDueDates(ProgrammingExercise programmingExerciseBeforeUpdate, ProgrammingExercise updatedProgrammingExercise) {
+        if (Boolean.TRUE.equals(updatedProgrammingExercise.isAllowOfflineIde())) {
+            ZonedDateTime now = ZonedDateTime.now();
+
+            if (programmingExerciseBeforeUpdate.getDueDate() != null && programmingExerciseBeforeUpdate.getDueDate().isBefore(now)
+                    && (updatedProgrammingExercise.getDueDate() == null || updatedProgrammingExercise.getDueDate().isAfter(now))) {
+                // New due date allows students to continue working on exercise
+                unlockAllRepositories(programmingExerciseBeforeUpdate.getId());
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java:1113]: I think individual due dates can also be earlier, so we also need to check them here.

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java (hunk 2 above)
Comment (JohannesStoehr) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java:1101]: I prefer the additional variable

**Review by tobias-lippert — DISMISSED** [review]
Tested all the described scenarios on ts2. They all work as expected. Good job.

File: src/main/java/de/tum/in/www1/artemis/service/messaging/InstanceMessageReceiveService.java [hunk 4]
```diff
@@ -201,6 +210,26 @@ public void processLockAllRepositories(Long exerciseId) {
         programmingExerciseScheduleService.lockAllStudentRepositories(programmingExercise).run();
     }
 
+    public void processUnlockAllRepositoriesWithoutEarlierIndividualDueDate(Long exerciseId) {
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/messaging/InstanceMessageReceiveService.java:213]: 2x Codacy suggestions about missing method Javadoc

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java [hunk 5]
```diff
@@ -1073,4 +1074,68 @@ public void deleteTasksWithSolutionEntries(Long exerciseId) {
...
+
+        boolean lockedUnlockedRepos = handleRepoAccessRightChangesDueDates(programmingExerciseBeforeUpdate, updatedProgrammingExercise);
+        if (lockedUnlockedRepos) {
+            return;
+        }
+
+        handleRepoAccessRightChangesChangesOfflineIDE(programmingExerciseBeforeUpdate, updatedProgrammingExercise);
+    }
+
+    private boolean handleRepoAccessRightChangesDueDates(ProgrammingExercise programmingExerciseBeforeUpdate, ProgrammingExercise updatedProgrammingExercise) {
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java:1107]: A Javadoc should explain the meaning of the `boolean` return value even for a private method.

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java [hunk 6]
```diff
@@ -1073,4 +1074,68 @@ public void deleteTasksWithSolutionEntries(Long exerciseId) {
...
+                    && updatedProgrammingExercise.getDueDate() != null && updatedProgrammingExercise.getDueDate().isBefore(now)) {
+                // New due date forbids students to continue working on exercise, if their individual due date does not override the new due date
+                instanceMessageSendService.sendLockAllRepositoriesWithoutLaterIndividualDueDate(programmingExerciseBeforeUpdate.getId());
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private boolean handleRepoAccessRightChangesChangesOfflineIDE(ProgrammingExercise programmingExerciseBeforeUpdate, ProgrammingExercise updatedProgrammingExercise) {
```
[the remark this answers is not in the document]
Comment (JohannesStoehr) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java:1127]: Lock means wright access in this case. With the offline IDE they need it to push. Without the offline IDE, Artemis handles everything including the pushing, so the student don't need write access.

Comment added

**Comment by JohannesStoehr** [comment]
Only added java comments, manual reviews are still valid

**Review by sleiss — APPROVED** [review]
Code changes LGTM

**Comment by krusche** [comment]
@JohannesStoehr Ion a future PR we should probably notify the instructor in the user interface upon exercise change in case this will lead to lock changes, so that the instructor is aware of it. Could you please implement this in a follow-up?

**Pull request merged**